src/dmd/dtemplate.d

```diff
@@ -8173,3 +8173,61 @@ MATCH matchArg(TemplateParameter tp, Scope* sc, RootObject oarg, size_t i, Templ
     else
         assert(0);
 }
+
+
+/***********************************************
+ * Collect and print statistics on template instantiations.
+ */
+struct TemplateStats
+{
+    __gshared TemplateStats[void*] stats;
+
+    uint numInstantiations;     // number of instantiations of the template
+    uint uniqueInstantiations;  // number of unique instantiations of the template
+
+    /*******************************
+     * Add this instance
+     */
+    static void incInstance(TemplateDeclaration td)
+    {
+        if (!global.params.vtemplates)
+            return;
+        if (!td)
+            return;
+        auto ts = cast(void*)td in stats;
+        if (ts)
+            ++stats[cast(void*)td].numInstantiations;
+        else
+            stats[cast(void*)td] = TemplateStats(1, 0);
```

Comment:
You can avoid the double lookup:

```suggestion
        if (auto ts = cast(void*)td in stats)
            ++ts.numInstantiations;
        else
            stats[cast(void*)td] = TemplateStats(1, 0);
```

Reply:
In the future, this could be simplified to:
```suggestion
        stats.require(cast(void*) td).numInstantiations++;
```

After we bump the minimum frontend version used for bootstrapping to 2.082 (https://dlang.org/phobos/object#.require 
https://dlang.org/changelog/2.082.0.html#require_update)